libsigner/src/messages.rs

```diff
@@ -0,0 +1,1195 @@
+    }
+}
+
+impl From<&SignerMessage> for TypePrefix {
+    fn from(message: &SignerMessage) -> Self {
+        match message {
+            SignerMessage::Packet(_) => TypePrefix::Packet,
+            SignerMessage::BlockResponse(_) => TypePrefix::BlockResponse,
+        }
+    }
+}
+
+impl From<&Packet> for TypePrefix {
+    fn from(packet: &Packet) -> Self {
+        match packet.msg {
+            Message::DkgBegin(_) => TypePrefix::DkgBegin,
+            Message::DkgPrivateBegin(_) => TypePrefix::DkgPrivateBegin,
+            Message::DkgEndBegin(_) => TypePrefix::DkgEndBegin,
+            Message::DkgEnd(_) => TypePrefix::DkgEnd,
+            Message::DkgPublicShares(_) => TypePrefix::DkgPublicShares,
+            Message::DkgPrivateShares(_) => TypePrefix::DkgPrivateShares,
+            Message::NonceRequest(_) => TypePrefix::NonceRequest,
+            Message::NonceResponse(_) => TypePrefix::NonceResponse,
+            Message::SignatureShareRequest(_) => TypePrefix::SignatureShareRequest,
+            Message::SignatureShareResponse(_) => TypePrefix::SignatureShareResponse,
+        }
+    }
+}
+
+/// The messages being sent through the stacker db contracts
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub enum SignerMessage {
+    /// The signed/validated Nakamoto block for miners to observe
+    BlockResponse(BlockResponse),
+    /// DKG and Signing round data for other signers to observe
+    Packet(Packet),
+}
+
+impl StacksMessageCodec for SignerMessage {
+    fn consensus_serialize<W: Write>(&self, fd: &mut W) -> Result<(), CodecError> {
+        write_next(fd, &(TypePrefix::from(self) as u8))?;
+        match self {
+            SignerMessage::Packet(packet) => {
+                consensus_serialize_packet(fd, packet)?;
+            }
+            SignerMessage::BlockResponse(block_response) => {
+                write_next(fd, block_response)?;
+            }
+        };
+        Ok(())
+    }
+
+    fn consensus_deserialize<R: Read>(fd: &mut R) -> Result<Self, CodecError> {
+        let type_prefix_byte = read_next::<u8, _>(fd)?;
+        let type_prefix = TypePrefix::try_from(type_prefix_byte)?;
+        let message = match type_prefix {
+            TypePrefix::Packet => {
+                let packet = consensus_deserialize_packet(fd)?;
+                SignerMessage::Packet(packet)
+            }
+            TypePrefix::BlockResponse => {
+                let block_response = read_next::<BlockResponse, _>(fd)?;
+                SignerMessage::BlockResponse(block_response)
+            }
+            _ => {
+                return Err(CodecError::DeserializeError(format!(
+                    "Unknown signer message type prefix: {}",
+                    type_prefix_byte
+                )))
+            }
+        };
+        Ok(message)
+    }
+}
+
+fn consensus_serialize_scalar<W: Write>(fd: &mut W, scalar: &Scalar) -> Result<(), CodecError> {
```

Comment:
Each of these `consensus_[de]serialize_TYPE` functions should really be trait implementations of `StacksMessageCodec`.  You would create a trait extension for each `TYPE`, and then implement that extension trait for the `TYPE` as well as `StacksMessageCodec` for the extension.

For example, you'd have `trait ScalarExtensions` here, and you'd implement it for `Scalar`, and you'd implement `StacksMessageCodec` for `T: ScalarExtensions`.

Reply:
Ahhh I didn't think of this. I was like "Oh can't implement a trait for a object outside of its definition folder". Will try this work around!

Reply:
I cannot get the trait to work for StacksMessageCodec. I ended up using just a local trait definition StacksMessageCodecExtensions and implementing it for all the wsts objects. Attempts to then use `impl<T: StacksMessageCodecExtensions> StacksMessageCodec for T` all failed with complaints about generics not being allowed for external traits